plugins/inputs/influxdb_listener/influxdb_listener.go

```diff
@@ -0,0 +1,404 @@
+			return err
+		}
+	}
+	h.listener = listener
+	h.port = listener.Addr().(*net.TCPAddr).Port
+
+	go func() {
+		err = h.server.Serve(h.listener)
+		if err != http.ErrServerClosed {
+			h.log.Infof("Error serving HTTP on %s", h.ServiceAddress)
+		}
+	}()
+
+	h.log.Infof("Started HTTP listener service on %s", h.ServiceAddress)
+
+	return nil
+}
+
+// Stop cleans up all resources
+func (h *InfluxDBListener) Stop() {
+	err := h.server.Shutdown(context.Background())
+	if err != nil {
+		h.log.Infof("Error shutting down HTTP server: %v", err.Error())
+	}
+
+	h.log.Infof("Stopped HTTP listener service on %s", h.ServiceAddress)
+}
+
+func (h *InfluxDBListener) ServeHTTP(res http.ResponseWriter, req *http.Request) {
+	h.requestsRecv.Incr(1)
+	h.mux.ServeHTTP(res, req)
+	h.requestsServed.Incr(1)
+}
+
+func (h *InfluxDBListener) handleQuery() http.HandlerFunc {
+	return func(res http.ResponseWriter, req *http.Request) {
+		defer h.queriesServed.Incr(1)
+		// Deliver a dummy response to the query endpoint, as some InfluxDB
+		// clients test endpoint availability with a query
+		res.Header().Set("Content-Type", "application/json")
+		res.Header().Set("X-Influxdb-Version", "1.0")
+		res.WriteHeader(http.StatusOK)
+		res.Write([]byte("{\"results\":[]}"))
+	}
+}
+
+func (h *InfluxDBListener) handlePing() http.HandlerFunc {
+	return func(res http.ResponseWriter, req *http.Request) {
+		defer h.pingsServed.Incr(1)
+		verbose := req.URL.Query().Get("verbose")
+
+		// respond to ping requests
+		if verbose != "" && verbose != "0" && verbose != "false" {
+			res.WriteHeader(http.StatusOK)
+			b, _ := json.Marshal(map[string]string{"version": "1.0"}) // based on header set above
+			res.Write(b)
+		} else {
+			res.WriteHeader(http.StatusNoContent)
+		}
+	}
+}
+
+func (h *InfluxDBListener) handleDefault() http.HandlerFunc {
+	return func(res http.ResponseWriter, req *http.Request) {
+		defer h.notFoundsServed.Incr(1)
+		http.NotFound(res, req)
+	}
+}
+
+func (h *InfluxDBListener) handleWrite() http.HandlerFunc {
+	return func(res http.ResponseWriter, req *http.Request) {
+		defer h.writesServed.Incr(1)
+		// Check that the content length is not too large for us to handle.
+		if req.ContentLength > h.MaxBodySize.Size {
+			tooLarge(res)
+			return
+		}
+
+		db := req.URL.Query().Get("db")
+
+		body := req.Body
+		body = http.MaxBytesReader(res, body, h.MaxBodySize.Size)
+		// Handle gzip request bodies
+		if req.Header.Get("Content-Encoding") == "gzip" {
+			var err error
+			body, err = gzip.NewReader(body)
+			if err != nil {
+				h.log.Debugf("Error decompressing request body: %v", err.Error())
+				badRequest(res, err.Error())
+				return
+			}
+			defer body.Close()
+		}
+
+		parser := influx.NewStreamParser(body)
+		parser.SetTimeFunc(h.timeFunc)
+
+		precisionStr := req.URL.Query().Get("precision")
+		if precisionStr != "" {
+			var precision time.Duration
+			precision = getPrecisionMultiplier(precisionStr)
+			parser.SetTimePrecision(precision)
```

Comment:
I still want to remove this, and here in this plugin multiple the metric as needed.

Reply:
We paired on this after your last comment and we realized that since the precision logic needs to know whether or not the timestamp is present in the line protocol, it makes to have that logic in MetricHandler.  Moving all the precision logic to the input plugin would require us to add a way to have the parser tell the plugin whether there is a timestamp which didn't seem worth it.

We were able to clean things up a little by removing metric.Builder and moving the timestamp truncation into MetricHandler.